homeassistant/components/habitica/__init__.py

```diff
@@ -147,7 +125,49 @@ async def handle_api_call(call):
             EVENT_API_CALL_SUCCESS, {"name": name, "path": path, "data": data}
         )
 
-    hass.services.async_register(
-        DOMAIN, SERVICE_API_CALL, handle_api_call, schema=SERVICE_API_CALL_SCHEMA
-    )
+    data = hass.data.setdefault(DOMAIN, {})
+    config = config_entry.data
+    websession = async_get_clientsession(hass)
+    url = config[CONF_URL]
+    username = config[CONF_API_USER]
+    password = config[CONF_API_KEY]
+    name = config.get(CONF_NAME)
+    config_dict = {"url": url, "login": username, "password": password}
+    api = HAHabitipyAsync(config_dict)
+    user = await api.user.get()
+    if name is None:
+        name = user["profile"]["name"]
+        hass.config_entries.async_update_entry(
+            config_entry,
+            data={**config_entry.data, CONF_NAME: name},
+        )
+    data[config_entry.entry_id] = api
+
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(config_entry, component)
+        )
+
+    if not hass.services.has_service(DOMAIN, SERVICE_API_CALL):
+        hass.services.async_register(
+            DOMAIN, SERVICE_API_CALL, handle_api_call, schema=SERVICE_API_CALL_SCHEMA
+        )
+
     return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    hass.components.webhook.async_unregister(SERVICE_API_CALL)
```

Comment:
Only unregister the service if it's the last config entry that is unloaded. We can use `hass.config_entries.async_entries` to get a list of all config entries for this domain.

Reply:
@MartinHjelmare, thanks again for all your help.
I have pushed the changes, thou am unsure if I should have simply checked for one entry here and if so, only then unregister the service.
Most of these concept are new for me, so I'm learning a lot, thank you